### PR TITLE
feat:H1083-L1-L3-L4-L5 [Payment Authorise and Sync] + [Refunds and Sync] + [Redirection Flow (BNPL)] + [3DS Payment and Sync] for MultiSafePay

### PR DESCRIPTION
**Pull request by zaid-ali753**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Implemented
`L1 - Payment Authorise and Sync`
`L3 - Refunds and Sync`
`L4 - Redirection Flow (BNPL)`
`L5 - 3DS Payment and Sync`

Not supported from Connector
`L2 - Manual Capture and Sync` 
`L6 - Mandate Flow`

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Tested via. [Postman](https://drive.google.com/file/d/1M-Vu2XUsg8gsOgW25Du-vNTom4XDDun5/view?usp=sharing)

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct CheckoutOptions {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub validate_cart: Option<bool>,
+    pub tax_tables: TaxObject
+}
+
+//TODO: Fill the struct with respective fields
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:212]: remove this todo

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+            _ => String::from("redirect"),
+        };
+
+        let gateway = match _item.request.payment_method_data {
+            api::PaymentMethod::Card(ref ccard) => Gateway::Amex,
+            api::PaymentMethod::PayLater(ref paylater) => Gateway::Klarna,
+            _ => Gateway::Amex,
+        };
+
+        println!("Printing Gatwaya ---> {:?}",gateway_info);
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:377]: remove this println

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+            days_active: Some(30),
+            seconds_active: Some(259200),
+            var1: None,
+            var2: None,
+            var3: None
+        })
+    }
+}
+
+//TODO: Fill the struct with respective fields
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:419]: remove this todo

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+            Ok(Self {
+                api_key: api_key.to_string(),
+            })
+        } else {
+            Err(errors::ConnectorError::FailedToObtainAuthType.into())
+        }
+    }
+}
+// PaymentsResponse
+//TODO: Append the remaining status flags
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:438]: remove this todo

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+    pub order_id: String,
+    pub currency: String,
+    pub amount:i64,
+    pub description: String,
+    pub capture: Option<String>,
+    pub payment_url: Option<Url>,
+    pub status: MultisafepayPaymentStatus,
+}
+
+//TODO: Fill the struct with respective fields
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:471]: remove this todo

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+                mandate_reference: None,
+                connector_metadata: None,
+            }),
+            amount_captured: Some(item.response.data.amount),
+            ..item.data
+        })
+    }
+}
+
+//TODO: Fill the struct with respective fields
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:520]: remove this todo

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+    Processing,
+}
+
+impl From<RefundStatus> for enums::RefundStatus {
+    fn from(item: RefundStatus) -> Self {
+        match item {
+            RefundStatus::Succeeded => Self::Success,
+            RefundStatus::Failed => Self::Failure,
+            RefundStatus::Processing => Self::Pending,
+            //TODO: Review mapping
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:569]: remove todo

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct MultisafepayPaymentsResponse {
+    pub success: bool,
+    pub data: Data,
+}
+
+impl<F,T> TryFrom<types::ResponseRouterData<F, MultisafepayPaymentsResponse, T, types::PaymentsResponseData>> for types::RouterData<F, T, types::PaymentsResponseData> {
+    type Error = error_stack::Report<errors::ParsingError>;
+    fn try_from(item: types::ResponseRouterData<F, MultisafepayPaymentsResponse, T, types::PaymentsResponseData>) -> Result<Self,Self::Error> {
+        println!("payment URL check {:?}", item.response.data.payment_url);
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:481]: remove this println

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+        println!("order_id check {:?}", item.response.data.order_id);
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:482]: remove this println

File: crates/router/src/connector/multisafepay/transformers.rs [hunk 10]
```diff
@@ -0,0 +1,612 @@
...
+            response: Ok(types::RefundsResponseData {
+                connector_refund_id: _item.response.data.refund_id.to_string(),
+                refund_status: enums::RefundStatus::from(refund_stat),
+            }),
+            .._item.data
+        })
+    }
+ }
+
+//TODO: Fill the struct with respective fields
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay/transformers.rs:624]: remove this todo

File: crates/router/src/services/api.rs [hunk 11]
```diff
@@ -163,7 +163,9 @@ where
         payments::CallConnectorAction::Trigger => {
             match connector_integration.build_request(req, &state.conf.connectors)? {
                 Some(request) => {
+                    println!("multisafepayreq --->{:?}",request);
```
Comment (jagan-jaya) [inline crates/router/src/services/api.rs:166]: remove this println

File: crates/router/src/services/api.rs [hunk 12]
```diff
@@ -163,7 +163,9 @@ where
...
                     let response = call_connector_api(state, request).await;
+                    println!("multisafepay --->{:?}",response);
```
Comment (jagan-jaya) [inline crates/router/src/services/api.rs:168]: remove this println

File: crates/router/tests/connectors/connector_auth.rs [hunk 13]
```diff
@@ -22,7 +23,7 @@ impl ConnectorAuthentication {
     pub(crate) fn new() -> Self {
         #[allow(clippy::expect_used)]
         toml::from_str(
-            &std::fs::read_to_string("tests/connectors/auth.toml")
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/connector_auth.rs:25]: undo this change

File: crates/router/tests/connectors/sample_auth.toml [hunk 14]
```diff
@@ -47,4 +47,7 @@ api_secret = "MySecretKey"
 [worldline]
 key1 = "Merchant Id"
 api_key = "API Key"
-api_secret = "API Secret Key"
\ No newline at end of file
+api_secret = "API Secret Key"
+
+[multisafepay]
+api_key = "b4f915f0ff94187f4e783cdc187354c81927c6ac"
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/sample_auth.toml:53]: remove this API key

File: crates/router/src/connector/multisafepay.rs [hunk 15]
```diff
@@ -0,0 +1,500 @@
...
+
+    fn get_url(
+        &self,
+        _req: &types::PaymentsSyncRouterData,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<String, errors::ConnectorError> {
+        let url = self.base_url(_connectors);
+        let mut api_key = self.get_auth_header(&_req.connector_auth_type)?[0].1.clone();
+        let ord_id = _req.payment_id.clone();
+        println!("Sync API being hit");
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay.rs:114]: remove this println

File: crates/router/src/connector/multisafepay.rs [hunk 16]
```diff
@@ -0,0 +1,500 @@
...
+            .method(services::Method::Post)
+            .url(&types::PaymentsAuthorizeType::get_url(
+                self, req, connectors,
+            )?)
+            .headers(types::PaymentsAuthorizeType::get_headers(
+                self, req, connectors,
+            )?)
+            .body(types::PaymentsAuthorizeType::get_request_body(self, req)?)
+            .build();
+        println!("authorizePay --->{:?}", requeue);
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay.rs:301]: remove this println

File: crates/router/src/connector/multisafepay.rs [hunk 17]
```diff
@@ -0,0 +1,500 @@
...
+            "{}v1/json/orders/{}/refunds?api_key={}",
+            url,
+            ord_id,
+            api_key
+        ))
+    }
+
+    fn get_request_body(&self, req: &types::RefundsRouterData<api::Execute>) -> CustomResult<Option<String>,errors::ConnectorError> {
+        let multisafepay_req = utils::Encode::<multisafepay::MultisafepayRefundRequest>::convert_and_encode(req).change_context(errors::ConnectorError::RequestEncodingFailed)?;
+        println!("refund_request_check --->{:?}",multisafepay_req);
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay.rs:369]: remove this println

File: crates/router/src/connector/multisafepay.rs [hunk 18]
```diff
@@ -0,0 +1,500 @@
...
+        let url = self.base_url(_connectors);
+        let mut api_key = self.get_auth_header(&_req.connector_auth_type)?[0].1.clone();
+        let ord_id = _req.payment_id.clone();
+        let x = format!(
+            "{}v1/json/orders/{}/refunds?api_key={}",
+            url,
+            ord_id,
+            api_key
+        );
+        print!("refund data check{:?}",x);
```
Comment (jagan-jaya) [inline crates/router/src/connector/multisafepay.rs:424]: remove this println

File: crates/router/tests/connectors/multisafepay.rs [hunk 19]
```diff
@@ -0,0 +1,438 @@
...
+        "multisafepay".to_string()
+    }
+}
+
+static CONNECTOR: MultisafepayTest = MultisafepayTest {};
+
+// Cards Positive Tests
+// Creates a payment using the manual capture flow (Non 3DS).
+#[actix_web::test]
+async fn should_only_authorize_payment() {
```
Comment (jagan-jaya) [inline crates/router/tests/connectors/multisafepay.rs:40]: please check if the tests are passing and update the PR description with test status

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/tests/connectors/multisafepay.rs (hunk 19 above)
Comment (zaid-ali753) [reply crates/router/tests/connectors/multisafepay.rs:40]: The connector does not supports Non 3DS flows and, capture and mandate are supported upon request to connector, so most unit tests will fail. But tested via. Postman. Adding link of [screenshots](https://drive.google.com/file/d/1M-Vu2XUsg8gsOgW25Du-vNTom4XDDun5/view?usp=sharing).
